terraform/ecs/main.tf

```diff
@@ -58,21 +58,120 @@ provider "aws" {
+resource "aws_ecs_cluster" "ecscluster" {
+  name = module.common.testing_id
+
+  depends_on = [
+    null_resource.iam_wait
+  ]
+}
+
+resource "aws_ecs_cluster_capacity_providers" "clustercapacity" {
+  cluster_name = aws_ecs_cluster.ecscluster.name
+
+  capacity_providers = [aws_ecs_capacity_provider.capacityprovider.name]
+
+}
+
+resource "aws_ecs_capacity_provider" "capacityprovider" {
+  name = "capacityprovider"
+
+  auto_scaling_group_provider {
+    auto_scaling_group_arn = aws_autoscaling_group.clusterasg.arn
+  }
+}
+
+data "aws_ami" "amazon_linux_2" {
+  most_recent = true
+  owners      = ["amazon"]
+
+  filter {
+    name   = "name"
+    values = ["amzn2-ami-ecs-hvm-*-x86_64-ebs"]
+  }
+}
+
+resource "aws_launch_template" "launchtemp" {
+  name_prefix   = "launchtemp"
+  image_id      = data.aws_ami.amazon_linux_2.image_id
+  instance_type = "t2.medium"
+  user_data     = base64encode("#!/bin/bash\necho ECS_CLUSTER=${module.common.testing_id} >> /etc/ecs/ecs.config")
+  metadata_options {
+    http_endpoint               = "enabled"
+    http_tokens                 = "required"
+    http_put_response_hop_limit = 2
+    instance_metadata_tags      = "enabled"
+  }
+  network_interfaces {
+    associate_public_ip_address = true
+    security_groups             = [module.basic_components.aoc_security_group_id]
+  }
+  iam_instance_profile {
+    name = aws_iam_instance_profile.cluster.name
+  }
+  depends_on = [
+    null_resource.iam_wait
+  ]
+}
+
+resource "aws_autoscaling_group" "clusterasg" {
+  name                = "clusterasg"
+  vpc_zone_identifier = module.basic_components.aoc_private_subnet_ids
+  desired_capacity    = 1
+  max_size            = 10
+  min_size            = 1
+
+  launch_template {
+    id      = aws_launch_template.launchtemp.id
+    version = "$Latest"
+  }
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+data "aws_iam_policy_document" "ecs_agent" {
+  statement {
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+  }
+}
+
+resource "aws_iam_role" "cluster_instance_role" {
+  assume_role_policy = data.aws_iam_policy_document.ecs_agent.json
+
+}
+
+resource "aws_iam_policy" "cluster_instance_policy" {
+  policy = file("instance-policy.json")
+}
+
+resource "aws_iam_policy_attachment" "cluster_instance_policy_attachment" {
+  name = "clusterpolicyattach"
+
+  roles      = [aws_iam_role.cluster_instance_role.id]
+  policy_arn = aws_iam_policy.cluster_instance_policy.arn
+}
+
+resource "aws_iam_instance_profile" "cluster" {
+  role = aws_iam_role.cluster_instance_role.name
+}
+
+resource "null_resource" "iam_wait" {
+  depends_on = [
+    aws_iam_role.cluster_instance_role,
+    aws_iam_policy.cluster_instance_policy,
+    aws_iam_policy_attachment.cluster_instance_policy_attachment,
+    aws_iam_instance_profile.cluster,
+  ]
+
+  provisioner "local-exec" {
+    command = "sleep 30"
+  }
```

Comment:
I understand what you are doing here with this `null_resource` but is the sleep needed? You should be able to depend on `iam_wait` without the sleep. It will only be complete after it's dependencies are resolved. If a provisioner is required I suggest just using something like `echo iam_wait done`.

Reply:
I added the sleep there because the previous module we used put a `sleep` there and I was trying to replicate it but I also think an echo would do the same and verified it in my local testing.  Changed in my latest commit.